### PR TITLE
Support for a new annotation @WithExecutor to use with @Asynchronous in FT

### PR DESCRIPTION
**Pull request by spericas**
### Description

Support for a new annotation @WithExecutor to override the default executor service to use when making an asynchronous call. Uses CDI to find a provider based on this annotation. Issue #8591.

### Documentation

Documentation will be provided in a follow-up PR.

### CODE REVIEW

File: microprofile/fault-tolerance/src/test/java/io/helidon/microprofile/faulttolerance/AsynchronousBean.java [hunk 1]
```diff
@@ -165,4 +169,34 @@ CompletableFuture<String> asyncCompletableFutureWithFallbackFailure() {
...
+        return Executors.newFixedThreadPool(10);
+    }
+
+    /**
+     * Bad executor name, should throw a {@link FaultToleranceException}.
+     *
+     * @return A future.
+     */
+    @Asynchronous
+    @WithExecutor("does-not-exist")
```
Comment (tomas-langer) [inline microprofile/fault-tolerance/src/test/java/io/helidon/microprofile/faulttolerance/AsynchronousBean.java:198]: Throwing an exception at runtime is not very user friendly.
Is there a chance to validate this during startup? Or maybe have an option to fallback to the default executor if the named one is not defined?

Comment (spericas) [reply microprofile/fault-tolerance/src/test/java/io/helidon/microprofile/faulttolerance/AsynchronousBean.java:198]: We could validate at startup I suppose. Silently falling back to the default one does not seem like a good idea to me when the task may depend on the right executor being used.

File: microprofile/fault-tolerance/src/main/java/io/helidon/microprofile/faulttolerance/WithExecutor.java [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Annotation used to specify an executor for an asynchronous call. Can be used
+ * e.g. to create platform instead of virtual threads. Targets only a method.
+ */
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD})
+@Inherited
+@Qualifier
+@InterceptorBinding
```
Comment (ljnelson) [inline microprofile/fault-tolerance/src/main/java/io/helidon/microprofile/faulttolerance/WithExecutor.java:37]: Just double-checking that you want this to be both a qualifier and an interceptor binding. (That's normally weird, though permitted.)

File: microprofile/fault-tolerance/src/main/java/io/helidon/microprofile/faulttolerance/MethodInvoker.java [hunk 3]
```diff
@@ -334,7 +336,17 @@ private CompletableFuture<Object> callSupplierNewThread(FtSupplier<Object> suppl
 
         // Call supplier in new thread
         ClassLoader ccl = Thread.currentThread().getContextClassLoader();
-        CompletableFuture<Object> asyncFuture = Async.create().invoke(() -> {
+        AsyncConfig.Builder asyncBuilder = AsyncConfig.builder();
+
+        // Handle a user-provided executor via @WithExecutor
+        if (introspector.hasWithExecutor()) {
+            WithExecutor withExecutor = introspector.withExecutor();
+            ExecutorService executorService = CDI.current().select(ExecutorService.class, withExecutor).get();
```
Comment (ljnelson) [inline microprofile/fault-tolerance/src/main/java/io/helidon/microprofile/faulttolerance/MethodInvoker.java:344]: (Per my earlier comment: I see here you are using it as a qualifier, so indeed this confirms that `@Qualifier` is certainly needed.)

**Comment by danielkec** [comment]
I like it! Can we also add a build-in executor which would be default for `Asynchronous`?

Or something like that? 
```java
@WithExecutor(FtExecutors.PLATFORM)
```
Would be great if such executor would be configurable.

**Comment by danielkec** [comment]
Also shutting down all executors with our qualifier on container shutdown would be nice

**Comment by spericas** [comment]
> Can we also add a build-in executor which would be default for `Asynchronous`?

 The default is now defined in `io.helidon.faulttolerance.FaultTolerance` and shared between SE and MP. Seems like the right default to me.

> 
> Or something like that?
> 
> ```java
> @WithExecutor(FtExecutors.PLATFORM)
> ```
> 
> Would be great if such executor would be configurable.

We already have `ThreadPoolSupplier` as a means to configure and create an `ExecutorService`. We could define a qualifier and add a built-in CDI provider for this. However, it's only few lines of code for anyone to create such a provider, and that could be easily documented too. Is it really worth it?

**Comment by spericas** [comment]
@danielkec I updated the main test to show how to use a configurable thread pool supplier to return an executor service for async. This could be easily documented.

**Comment by spericas** [comment]
After discussing the use case further with @danielkec, we agreed to proceed with a two-pronged approach: (i) an extension for the FT module to support using a different executor service (this PR!) and (ii) a new, simpler annotation in another module to only handle running arbitrary methods in platform threads. There will be a new issue created for (ii).

File: microprofile/fault-tolerance/src/main/java/io/helidon/microprofile/faulttolerance/FaultToleranceExtension.java [hunk 4]
```diff
@@ -230,11 +232,25 @@ void validateAnnotations(BeanManager bm,
...
             });
         }
     }
 
+    static void validateWithExecutor(BeanManager bm, Method method) {
+        WithExecutor withExecutor = method.getAnnotation(WithExecutor.class);
+        if (withExecutor != null) {
+            Set<Bean<?>> beans = bm.getBeans(ExecutorService.class, withExecutor);
+            if (beans.isEmpty()) {
+                throw new FaultToleranceDefinitionException("Unable to resolved named executor service '"
```
Comment (danielkec) [inline microprofile/fault-tolerance/src/main/java/io/helidon/microprofile/faulttolerance/FaultToleranceExtension.java:246]:
```suggestion
                throw new FaultToleranceDefinitionException("Unable to resolve named executor service '"
```

File: microprofile/fault-tolerance/src/test/java/io/helidon/microprofile/faulttolerance/AsynchronousTest.java [hunk 5]
```diff
@@ -123,4 +122,12 @@ void testAsyncCompletableFutureWithFallbackFailure() throws Exception {
         assertThat(completableFuture.get(), is("fallback"));
         assertThat(bean.wasCalled(), is(true));
     }
+
+    @Test
+    void testAsyncWithExecutor() throws Exception {
+        assertThat(bean.wasCalled(), is(false));
+        CompletableFuture<String> future = bean.asyncWithExecutor();
+        future.get();
```
Comment (danielkec) [inline microprofile/fault-tolerance/src/test/java/io/helidon/microprofile/faulttolerance/AsynchronousTest.java:130]: timeout

File: microprofile/fault-tolerance/src/test/java/io/helidon/microprofile/faulttolerance/AsynchronousBean.java [hunk 6]
```diff
@@ -165,4 +172,47 @@ CompletableFuture<String> asyncCompletableFutureWithFallbackFailure() {
...
+     * @return a future
+     * @see ExecutorServiceProducer#threadPoolSupplier()
+     */
+    @Asynchronous
+    @WithExecutor("platform-thread-executor")
+    CompletableFuture<String> asyncWithExecutor() {
+        Thread thread = Thread.currentThread();
+        called.set(thread.getName().contains("platform-threads") && !thread.isVirtual());
+        FaultToleranceTest.printStatus("AsynchronousBean::asyncWithExecutor", "success");
+        return CompletableFuture.completedFuture("success");
```
Comment (danielkec) [inline microprofile/fault-tolerance/src/test/java/io/helidon/microprofile/faulttolerance/AsynchronousBean.java:189]: Why not return the thread ref and doing the assertion in the test method? May be more clear what is wrong if the test fails

File: microprofile/fault-tolerance/src/main/java/io/helidon/microprofile/faulttolerance/FaultToleranceExtension.java (hunk 4 above)
Comment (spericas) [reply microprofile/fault-tolerance/src/main/java/io/helidon/microprofile/faulttolerance/FaultToleranceExtension.java]: Fixed!

File: microprofile/fault-tolerance/src/test/java/io/helidon/microprofile/faulttolerance/AsynchronousTest.java (hunk 5 above)
Comment (spericas) [reply microprofile/fault-tolerance/src/test/java/io/helidon/microprofile/faulttolerance/AsynchronousTest.java:130]: All these tests need to be "fixed" in that way. The implementation "should" never block that way.

File: microprofile/fault-tolerance/src/test/java/io/helidon/microprofile/faulttolerance/AsynchronousBean.java (hunk 6 above)
Comment (spericas) [reply microprofile/fault-tolerance/src/test/java/io/helidon/microprofile/faulttolerance/AsynchronousBean.java:189]: Again, all these tests can be improved, but that's for another PR.

**Pull request closed without merging**